docs/manual/inference.tex

```diff
@@ -126,22 +84,209 @@
+source code, but they do not run your program.
 
 
 \sectionAndLabel{Whole-program inference}{whole-program-inference}
 
-Whole-program inference is an interprocedural inference that
+Whole-program inference
 infers types for fields, method parameters, and method return types that do not
 have a user-written qualifier (for the given type system).
-The inferred type qualifiers are inserted into
-your program.
+The inferred type qualifiers are output into annotation files.
 The inferred type is the most specific type that is compatible with all the
 uses in the program.  For example, the inferred type for a field is the
 least upper bound of the types of all the expressions that are assigned
 into the field.
 
+There are three scripts that you can use to run whole-program inference.
+Each has advantages and disadvantages, discussed below:
+
+\begin{itemize}
+    \item
+    To run whole-program inference on a single project without modifying its source code,
+    use the \<wpi.sh> script (Section~\ref{wpi-one}). This script can automatically understand
+    many Ant, Maven, and Gradle build files, so it requires little manual configuration.
+
+    \item
+    To run whole-program inference on many projects without modifying their source code
+    (say, when running it on projects from GitHub), use the \<wpi-many.sh> script (Section~\ref{wpi-many}).
+    This script can understand the same build files as \<wpi.sh>.
+
+    \item
+    If you want to insert the inferred annotations directly into a single
+    project's source code, use the \<infer-and-annotate.sh> script (Section~\ref{wpi-insert}).
+\end{itemize}
+
+These type inference scripts appear in the \<checker/bin/> directory.
+The remainder of this chapter describes them
+(Sections~\ref{wpi-one}--\ref{wpi-insert}), then concludes with discussion
+that applies to all of them.x
+
+
+\sectionAndLabel{Running whole-program inference on a single project}{wpi-one}
+
+A typical invocation of \<wpi.sh> is
+
+\begin{Verbatim}
+  wpi.sh -- --checker nullness
+\end{Verbatim}
+
+The result is a set of log files placed in the \<dljc-out> folder of the
+target project. The results of type-checking with each candidate set of
+annotations will be concatenated into the file \<dljc-out/wpi.log>.
+
+The full syntax for invoking \<wpi.sh> is
+
+\begin{Verbatim}
+  wpi.sh [-d PROJECTDIR] [-t TIMEOUT] -- [DLJC-ARGS]
+\end{Verbatim}
+
+Arguments in square brackets are optional.
+Here is an explanation of the arguments:
+
+\begin{description}
+\item[-d PROJECTDIR]
+  The top-level directory of the project.  It must contain an Ant, Gradle,
+  or Maven buildfile. The default is the current working directory.
+
+\item[-t TIMEOUT]
+  The timeout for running the checker, in seconds
+
+\item[DLJC-ARGS]
+  Arguments that are passed directly to
+  \ahref{https://github.com/kelloggm/do-like-javac}{do-like-javac}'s
+  \<dljc> program without
+  modification.  One argument is required: \<-\relax-checker>, which indicates
```

Comment:
https://github.com/kelloggm/do-like-javac says:

> The --checker option is required. Its argument is the fully-qualified name of the checker you want to run.

Can `--checker` be supplied multiple times?  Or can I supply multiple checkers in comma-separated form, as in javac's -processor argument (https://checkerframework.org/manual/#shorthand-for-checkers)?

(Please fix that in do-like-javac rather than here, though this text might need to be tweaked slightly.)